website/api/views.py

```diff
@@ -649,3 +654,95 @@ class CompanyViewSet(viewsets.ModelViewSet):
+        data["slug"] = slug
+
+        contributors = Project.get_contributors(self, data["github_url"])  # get contributors
+
+        project_instance = Project(
+            name=data["name"],
+            slug=slug,
+            description=data.get("description", ""),
+            github_url=data.get("github_url", ""),
+            wiki_url=data.get("wiki_url", ""),
+            homepage_url=data.get("homepage_url", ""),
+            logo=data.get("logo", ""),
+        )
+
+        # If the logo is not provided get the logo from the repo itself
+        if project_instance.logo == "":
+            logo = project_instance.get_github_logo_save_to_storage(project_instance.github_url)
+            if logo:
+                project_instance.logo = logo
+
+        project_instance.save()
+
+        # Set contributors
+        if contributors:
+            project_instance.contributors.set(contributors)
+
+        serializer = ProjectSerializer(project_instance)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    def list(self, request, *args, **kwargs):
+        projects = Project.objects.prefetch_related("contributors").all()
+
+        project_data = []
+        for project in projects:
+            contributors_data = []
+            for contributor in project.contributors.all():
+                contributor_info = ContributorSerializer(contributor)
+                contributors_data.append(contributor_info.data)
+            contributors_data.sort(key=lambda x: x["contributions"], reverse=True)
+            project_info = ProjectSerializer(project)
+            project_info = {
+                "id": project.id,
+                "name": project.name,
+                "slug": project.slug,
+                "description": project.description,
+                "github_url": project.github_url,
+                "wiki_url": project.wiki_url,
+                "homepage_url": project.homepage_url,
+                "logo": project.logo.url if project.logo else None,
+                "created": project.created,
+                "modified": project.modified,
+                "contributors": contributors_data,
+            }
+            project_data.append(project_info)
+
+        return Response(
+            {"count": len(project_data), "projects": project_data},
+            status=200,
+        )
+
+    def update(self, request, *args, **kwargs):
+        projects = Project.objects.prefetch_related("contributors").all()
+        for project in projects:
+            print(project.github_url)
+            contributors = Project.get_contributors(self, github_url=project.github_url)
+            project.contributors.set(contributors)
+        serializer = ProjectSerializer(projects, many=True)
+        return Response(
+            {"count": len(projects), "projects": serializer.data}, status=status.HTTP_200_OK
+        )
```

Comment:
This method's behavior isn't following normal REST API design. I'd expect this to be responsible for a single `Project` instance update.

Reply:
How should i approach this? a new function that works for a single project ```upate_project``` and call that for each project in another function ```update```?

Reply:
I'd start from the beginning -- why would a user need this endpoint? You could use a periodic sync job for that, e.g. Django command.

If you sure you need it to be available via endpoint I suggest using a separate scope for that and remove it from the `ProjectViewSet`

Reply:
Made this admin only as it should be... hope it resolves this..